File: clean.pl
#!/usr/bin/perl
use strict;
use warnings;
use File::Spec;
use File::Basename;

# Directory containing the router configurations
my $input_dir = File::Spec->catdir('..', 'CSR', 'CSR', 'in-situ-configs');
my $output_dir = 'sanitized_configs';

# Create output directory if it doesn't exist
mkdir $output_dir unless -d $output_dir;

# Define rotation key (e.g., "2+,4-,3-,6+")
my @rotation_key = ( ['2', '+'], ['4', '-'], ['3', '-'], ['6', '+'] );

# Function to rotate an octet (8-bit integer)
sub rotate_octet {
    my ($octet, $bits, $direction) = @_;
    
    return $octet if $octet < 0 || $octet > 255;  # Sanity check

    if ($direction eq "+") {
        # Rotate right
        return (($octet >> $bits) | ($octet << (8 - $bits))) & 0xFF;
    } else {
        # Rotate left
        return (($octet << $bits) | ($octet >> (8 - $bits))) & 0xFF;
    }
}

# Function to sanitize IP addresses
sub sanitize_ip {
    my ($ip, $log_fh) = @_;
    my @octets = split(/\./, $ip);

    for my $i (0..3) {
        my ($bits, $direction) = @{$rotation_key[$i]};
        $octets[$i] = rotate_octet($octets[$i], $bits, $direction);
    }

    return join('.', @octets);
}

# Function to sanitize cryptographic text
sub sanitize_cryptotext {
    my ($text, $log_fh) = @_;
    # Fixed the regex here
    $text =~ s/(\$9\$\S\S)(\S{27})(\S)\S/$1$2_$3_/;
    return $text;
}

# Function to sanitize SNMP community strings
sub sanitize_snmp {
    my ($line, $log_fh) = @_;
    $line =~ s/(snmp community )([A-Za-z0-9_]+)/$1public/;
    return $line;
}

# Function to sanitize passwords
sub sanitize_password {
    my ($line, $log_fh) = @_;
    $line =~ s/(encrypted-password )(.)(.{10})(..);/$1X$2$3X;/;
    return $line;
}

# Process each file in the input directory
opendir(my $dh, $input_dir) || die "Can't open $input_dir: $!";
while (my $filename = readdir($dh)) {
    next if $filename =~ /^\./;  # Skip hidden files and directories

    my $input_file = File::Spec->catfile($input_dir, $filename);
    my $output_file = File::Spec->catfile($output_dir, "sanitized_$filename");
    my $log_file = File::Spec->catfile($output_dir, "log_$filename.log");

    open my $in_fh, '<', $input_file or die "Cannot open $input_file: $!";
    open my $out_fh, '>', $output_file or die "Cannot open $output_file: $!";
    open my $log_fh, '>', $log_file or die "Cannot open $log_file: $!";

    while (my $line = <$in_fh>) {
        chomp $line;

        # Sanitize IP addresses
        if ($line =~ /(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})/) {
            my $ip = $1;
            print $log_fh "Found IP address:        $ip\n";
            my $sanitized_ip = sanitize_ip($ip, $log_fh);
            print $log_fh "IP address sanitized to: $sanitized_ip\n";
            $line =~ s/\Q$ip\E/$sanitized_ip/g;
        }

        # Sanitize passwords
        if ($line =~ /(encrypted-password .+)/) {
            print $log_fh "Found encrypted-password field...\n";
            $line = sanitize_password($line, $log_fh);
        }

        # Sanitize cryptographic text (e.g., `$9$` hashes)
        if ($line =~ /(\$9\$\S{1}[A-Za-z0-9\/\.]{29})/) {
            my $cryptotext = $1;
            print $log_fh "Found cryptographic hash:     $cryptotext\n";
            my $sanitized_cryptotext = sanitize_cryptotext($cryptotext, $log_fh);
            print $log_fh "Crypto hash sanitized to:     $sanitized_cryptotext\n";
            $line =~ s/\Q$cryptotext\E/$sanitized_cryptotext/g;
        }

        # Sanitize SNMP community strings
        if ($line =~ /(snmp community .+)/) {
            print $log_fh "Found SNMP community string...\n";
            $line = sanitize_snmp($line, $log_fh);
        }

        # Output the sanitized line
        print $out_fh "$line\n";
    }

    close $in_fh;
    close $out_fh;
    close $log_fh;

    print "Processed $filename. Sanitized output written to $output_file and logs written to $log_file.\n";
}
closedir($dh);

print "All files processed. Sanitized configs are in the '$output_dir' directory.\n";

File: clean.py
#!/usr/bin/env python3
import sys
import os
import re
import argparse
from pathlib import Path

"""
clean6.py
Removes sensitive (sanitize) information from Juniper router configuration files
(c)2025 wynn.fenwick@telus.com
TELUS CONFIDENTIAL
"""

# Define rotation key (e.g., "2+,4-,3-,6+")
ROTATION_KEY = [('2', '+'), ('4', '-'), ('3', '-'), ('6', '+')]

def parse_arguments():
    parser = argparse.ArgumentParser(
        description='Removes sensitive (sanitize) information from Juniper router configuration files.')
    parser.add_argument('-m', action='store_true', help='Skip more prompts')
    parser.add_argument('-a', action='store_true', help='Rotate IP addresses (default: no rotation)')
    return parser.parse_args()

def rotate_octet(octet, bits, direction):
    """Rotate an octet (8-bit integer)"""
    try:
        octet = int(octet)
        if not 0 <= octet <= 255:
            return octet
        
        bits = int(bits)
        if direction == "+":
            # Rotate right
            return ((octet >> bits) | (octet << (8 - bits))) & 0xFF
        else:
            # Rotate left
            return ((octet << bits) | (octet >> (8 - bits))) & 0xFF
    except ValueError:
        return octet

def sanitize_ip(ip, args, log_fh):
    """Sanitize IP addresses"""
    octets = ip.split('.')
    
    if args.a:
        for i in range(4):
            bits, direction = ROTATION_KEY[i]
            octets[i] = str(rotate_octet(octets[i], bits, direction))
    
    return '.'.join(octets)

def sanitize_snmp(line, log_fh):
    """Sanitize SNMP community strings"""
    return re.sub(r'community "(\S+)" \{', 'community "OBFUSCATED" {', line)

def sanitize_password(line, log_fh):
    """Sanitize passwords and hashes"""
    patterns = [
        (r'(^\s+authentication-key "\$1\$)(..)([^;]+)(..)(\";.*)', r'\1XX\3XX\5'),
        (r'(^\s+encrypted-password "\$1\$)(..)([^;]+)(..)(\";.*)', r'\1XX\3XX\5'),
        (r'(^\s+secret "\$1\$)(..)([^;]+)(..)(\";.*)', r'\1XX\3XX\5'),
        (r'(^\s+authentication-key "\$9\$)(..)([^;]+)(..)(\";.*)', r'\1XX\3XX\5'),
        (r'(^\s+secret "\$9\$)(..)([^;]+)(..)(\";.*)', r'\1XX\3XX\5'),
        (r'(encrypted-password\s+)(..)([^;]+)(..)(;.*)', r'\1XX\3XX\5')
    ]
    
    for pattern, replacement in patterns:
        if re.search(pattern, line):
            return re.sub(pattern, replacement, line)
    
    return line

def process_input(in_fh, out_fh, log_fh, args):
    """Process a single input stream"""
    for line in in_fh:
        # Skip more prompts if -m option is set
        if args.m and re.search(r'---\(more( \d{1,2}%)?\)---', line):
            continue
            
        line = line.rstrip()
        
        # Sanitize IP addresses
        ip_match = re.search(r'(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})', line)
        if ip_match:
            ip = ip_match.group(1)
            print(f"Found IP address:        {ip}", file=log_fh)
            sanitized_ip = sanitize_ip(ip, args, log_fh)
            print(f"IP address sanitized to: {sanitized_ip}", file=log_fh)
            line = line.replace(ip, sanitized_ip)
        
        # Sanitize passwords and hashes
        if re.search(r'encrypted-password|(\$[19]\$)', line):
            line = sanitize_password(line, log_fh)
            print(f"Obfuscating password/hash field {line.lstrip()} ...", file=log_fh)
            line = line.replace("## SECRET-DATA", "## SECRET-DATA-OBFUSCATED")
        
        # Sanitize SNMP community strings
        if re.search(r'(community .+)', line):
            print("Found SNMP community string...", file=log_fh)
            line = sanitize_snmp(line, log_fh)
        
        # Output the sanitized line
        if re.search(r'\S', line):  # Only print lines with non-whitespace characters
            print(line, file=out_fh)

def main():
    args = parse_arguments()
    
    # Check if we're receiving input from stdin
    if not sys.stdin.isatty():
        # Processing from stdin
        with open('sanitize.log', 'w') as log_fh:
            process_input(sys.stdin, sys.stdout, log_fh, args)
        print("Sanitization complete. Logs written to sanitize.log.", file=sys.stderr)
    else:
        # Processing directory
        input_dir = Path('.') / 'CSR' / 'CSR' / 'in-situ-configs'
        output_dir = Path('sanitized_configs')
        
        # Create output directory if it doesn't exist
        output_dir.mkdir(exist_ok=True)
        
        # Process each file in the input directory
        for input_file in input_dir.glob('*'):
            if input_file.name.startswith('.'):
                continue
                
            output_file = output_dir / f"sanitized_{input_file.name}"
            log_file = output_dir / f"log_{input_file.name}.log"
            
            with open(input_file, 'r') as in_fh, \
                 open(output_file, 'w') as out_fh, \
                 open(log_file, 'w') as log_fh:
                process_input(in_fh, out_fh, log_fh, args)
                
            print(f"Processed {input_file.name}. Sanitized output written to {output_file} "
                  f"and logs written to {log_file}")
        
        print(f"All files processed. Sanitized configs are in the '{output_dir}' directory.")

if __name__ == '__main__':
    main()